app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheetControllerV2.kt

```diff
@@ -445,54 +452,73 @@ class BloodPressureEntrySheetControllerV2 @Inject constructor(
         .withLatestFrom(screenChanges)
         .filter { (_, screen) -> screen == DATE_ENTRY }
 
+    val prefilledDateStream = events
+        .ofType<BloodPressureDateToPrefillCalculated>()
+        .map { it.date }
+
     val newBpDataStream = Observables.combineLatest(dateValidations, bpValidations, patientUuidStream)
         .map { (dateResult, bpResult, patientUuid) ->
-          when {
-            dateResult is Valid && bpResult is Success -> SaveBpData.ReadyToCreate(dateResult.parsedDate, bpResult, patientUuid)
-            else -> SaveBpData.NeedsCorrection
+          if (dateResult is Valid && bpResult is Success) {
+            SaveBpData.ReadyToCreate(dateResult.parsedDate, bpResult.systolic, bpResult.diastolic, patientUuid)
+          } else {
+            SaveBpData.NeedsCorrection
           }
         }
 
     val updateBpDataStream = Observables.combineLatest(dateValidations, bpValidations, existingBpUuidStream)
         .map { (dateResult, bpResult, bpUuid) ->
-          when {
-            dateResult is Valid && bpResult is Success -> SaveBpData.ReadyToUpdate(dateResult.parsedDate, bpResult, bpUuid)
-            else -> SaveBpData.NeedsCorrection
+          if (dateResult is Valid && bpResult is Success) {
+            SaveBpData.ReadyToUpdate(dateResult.parsedDate, bpResult.systolic, bpResult.diastolic, bpUuid)
+          } else {
+            SaveBpData.NeedsCorrection
           }
         }
 
     val saveNewBp = saveClicks
         .withLatestFrom(newBpDataStream) { _, newBp -> newBp }
         .ofType<SaveBpData.ReadyToCreate>()
-        .flatMapSingle { (date, bp, patientUuid) ->
-          val dateAsInstant = date.atTime(OffsetTime.now(clock)).toInstant()
-          bloodPressureRepository.saveMeasurement(patientUuid, bp.systolic, bp.diastolic, dateAsInstant)
+        .withLatestFrom(prefilledDateStream)
+        .flatMapSingle { (newBp, prefilledDate) ->
+          val wasDateChanged = prefilledDate != newBp.date
+
+          bloodPressureRepository
+              .saveMeasurement(newBp.patientUuid, newBp.systolic, newBp.diastolic, newBp.date.toInstant(clock))
+              .map { BloodPressureSaved(wasDateChanged) }
```

Comment:
Instead of dismissing the sheet here, I added an intermediate event called `BloodPressureSaved`. Its analytics name is dependent on whether the date was changed or not so that we can track retro-active BP entries from Heap. 

The sheet is dismissed from another function called `dismissSheetWhenBpIsSaved()`